src/core/xfa/formcalc_lexer.js

```diff
@@ -0,0 +1,372 @@
+        if (data.charCodeAt(this.pos) === 0x22 /* = " */) {
+          // Escaped quote.
+          str.push(data.slice(start, this.pos++));
+          start = this.pos;
+          continue;
+        }
+        // End of string
+        break;
+      }
+
+      if (char === 0x5c /* = \ */) {
+        const match = data.substring(this.pos, this.pos + 10).match(hexPattern);
+        if (!match) {
+          continue;
+        }
+
+        str.push(data.slice(start, this.pos - 1));
+        const code = match[1];
+        if (code.length === 4) {
+          str.push(String.fromCharCode(parseInt(code, 16)));
+          start = this.pos += 5;
+        } else if (code.length !== 8) {
+          str.push(String.fromCharCode(parseInt(code.slice(0, 4), 16)));
+          start = this.pos += 5;
+        } else {
+          str.push(String.fromCharCode(parseInt(code, 16)));
+          start = this.pos += 9;
+        }
+      }
+    }
+
+    const lastChunk = data.slice(start, this.pos - 1);
+    if (str.length === 0) {
+      return new Token(TOKEN.string, lastChunk);
+    }
+
+    str.push(lastChunk);
+    const string = str.join("");
+    str.length = 0;
+
+    return new Token(TOKEN.string, string);
+  }
+
+  getNumber(first) {
+    const match = this.data.substring(this.pos).match(numberPattern);
+    if (!match) {
+      return first - 0x30 /* = 0 */;
+    }
+    const number = parseFloat(
+      this.data.substring(this.pos - 1, this.pos + match[0].length)
+    );
+
+    this.pos += match[0].length;
+    return new Token(TOKEN.number, number);
+  }
+
+  getCompOperator(alt1, alt2) {
+    if (this.data.charCodeAt(this.pos) === 0x3d /* = = */) {
+      this.pos++;
+      return new Token(alt1);
+    }
+    return new Token(alt2);
+  }
+
+  getLower() {
+    const char = this.data.charCodeAt(this.pos);
+    if (char === 0x3d /* = = */) {
+      this.pos++;
+      return new Token(TOKEN.le);
+    }
+
+    if (char === 0x3e /* = > */) {
+      this.pos++;
+      return new Token(TOKEN.ne);
+    }
+
+    return new Token(TOKEN.lt);
+  }
+
+  getSlash() {
+    if (this.data.charCodeAt(this.pos) === 0x2f /* = / */) {
+      this.skipUntilEOL();
+      return false;
+    }
+    return true;
+  }
+
+  getDot() {
+    const char = this.data.charCodeAt(this.pos);
+    if (char === 0x2e /* = . */) {
+      this.pos++;
+      return new Token(TOKEN.dotDot);
+    }
+
+    if (char === 0x2a /* = * */) {
+      this.pos++;
+      return new Token(TOKEN.dotStar);
+    }
+
+    if (char === 0x23 /* = # */) {
+      this.pos++;
+      return new Token(TOKEN.dotHash);
+    }
+
+    if (0x30 /* = 0 */ <= char && char <= 0x39 /* = 9 */) {
+      this.pos++;
+      const match = this.data.substring(this.pos).match(dotNumberPattern);
+      if (!match) {
+        return new Token(TOKEN.number, (char - 0x30) /* = 0 */ / 10);
+      }
+      const end = this.pos + match[0].length;
+      const number = parseFloat(this.data.substring(this.pos - 2, end));
```

Comment:
What should happen in this language if we parse a really large number that would be infinity in JS? e.g. "9e99999"

Reply:
Numbers are following IEEE754 format: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=1056
like in JS: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Numbers_and_dates
So using `parseFloat` should give the correct number.